Content.Client/Research/UI/RoboticsConsoleWindow.xaml.cs

```diff
@@ -0,0 +1,155 @@
+        };
+        DestroyButton.OnPressed += _ =>
+        {
+            OnDestroyPressed?.Invoke(_selected!);
+        };
+
+        // cant put multiple styles in xaml for some reason
+        DestroyButton.StyleClasses.Add(StyleBase.ButtonCaution);
+    }
+
+    public void UpdateState(RoboticsConsoleState state)
+    {
+        _cyborgs = state.Cyborgs;
+        _nextDestroy = state.NextDestroy;
+
+        // clear invalid selection
+        if (_selected is {} selected && !_cyborgs.ContainsKey(selected))
+            _selected = null;
+
+        var hasCyborgs = _cyborgs.Count > 0;
+        NoCyborgs.Visible = !hasCyborgs;
+        CyborgsContainer.Visible = hasCyborgs;
+        PopulateCyborgs();
+
+        PopulateData();
+
+        var access = _player.LocalPlayer?.ControlledEntity is {} player && _accessReader.IsAllowed(player, _console);
+        DangerZone.Visible = access;
+    }
+
+    private void PopulateCyborgs()
+    {
+        // _selected might get set to null when recreating so copy it first
+        var selected = _selected;
+        Cyborgs.Clear();
+        foreach (var (address, data) in _cyborgs)
+        {
+            var chassis = _proto.Index<EntityPrototype>(data.Chassis);
+            var icon = _sprite.GetPrototypeIcon(chassis).Default;
+            var item = Cyborgs.AddItem(data.Name, icon, metadata: address);
+            item.Selected = address == selected;
+        }
+        _selected = selected;
+    }
+
+    private void PopulateData()
+    {
+        if (_selected is not {} selected)
+        {
+            SelectCyborg.Visible = true;
+            BorgContainer.Visible = false;
+            return;
+        }
+
+        SelectCyborg.Visible = false;
+        BorgContainer.Visible = true;
+
+        var data = _cyborgs[selected];
+        var chassis = _proto.Index<EntityPrototype>(data.Chassis);
+        var model = chassis.Name;
+        var icon = _sprite.GetPrototypeIcon(chassis).Default;
+
+        BorgSprite.Texture = icon;
+
+        var batteryColor = data.Charge switch {
+            < 0.2f => "red",
+            < 0.4f => "orange",
+            < 0.6f => "yellow",
+            < 0.8f => "green",
+            _ => "blue"
+        };
+
+        var text = FormattedMessage.FromMarkup(Loc.GetString("robotics-console-model", ("name", model)));
+        text.PushNewline();
+        text.AddMarkup(Loc.GetString("robotics-console-designation"));
+        text.AddText($" {data.Name}\n"); // prevent players trolling by naming borg [color=red]satan[/color]
+        text.AddMarkup(Loc.GetString("robotics-console-battery", ("charge", (int) (data.Charge * 100f)), ("color", batteryColor)));
+        text.PushNewline();
+        text.AddMarkup(Loc.GetString("robotics-console-brain", ("brain", data.HasBrain)));
+        text.PushNewline();
+        text.AddMarkup(Loc.GetString("robotics-console-modules", ("count", data.ModuleCount)));
+        BorgInfo.SetMessage(text);
```

Comment:
Is this supposed to be pushmarkup

Reply:
idk what that is probably

Reply:
Just add it to formattedmessage on engine.

Reply:
https://github.com/space-wizards/RobustToolbox/pull/4924